src/graph/util/ExpressionUtils.cpp

```diff
@@ -536,7 +536,65 @@ StatusOr<Expression *> ExpressionUtils::foldConstantExpr(const Expression *expr)
     }
     return foldedExpr;
   }
-  return newExpr;
+
+  auto matcher = [](const Expression *e) {
+    return e->kind() == Expression::Kind::kLogicalAnd || e->kind() == Expression::Kind::kLogicalOr;
+  };
+  auto rewriter = [](const Expression *e) {
+    auto logicalExpr = static_cast<const LogicalExpression *>(e);
+    return simplifyLogicalExpr(logicalExpr);
+  };
+  return RewriteVisitor::transform(newExpr, matcher, rewriter);
+}
+
+Expression *ExpressionUtils::simplifyLogicalExpr(const LogicalExpression *logicalExpr) {
+  auto *expr = static_cast<LogicalExpression *>(logicalExpr->clone());
+  if (expr->kind() == Expression::Kind::kLogicalXor) return expr;
```

Comment:
the matcher has filtered only AND and OR, why check Xor here?

Reply:
This inspection is a double guarantee. And the matcher is not written in the function `simplifyLogicalExpr`.